scipy/sparse/linalg/tests/test_interface.py

```diff
@@ -214,6 +215,237 @@ def test_matmul(self):
         assert_raises(ValueError, operator.matmul, 2, A)
 
 
+class TestDotTests:
+    class OpArgs(NamedTuple):
+        ny: int
+        nx: int
+        complex: bool
+        dtype: str
+        
+    args1: OpArgs = OpArgs(11, 11, False, "int32")  # real integer square
+    args2: OpArgs = OpArgs(21, 11, False, "float64")  # overdetermined real
+    args1j: OpArgs = OpArgs(11, 11, True, "complex128")  # square complex
+    args2j: OpArgs = OpArgs(21, 11, True, "complex128")  # overdetermined complex
+    args3: OpArgs = OpArgs(11, 21, False, "float64")  # underdetermined real
+    
+    def dottest(
+        self,
+        Op: interface.LinearOperator,
+        rtol: float = 1e-6,
+        atol: float = 1e-21,
+        complexflag: bool = False,
+    ) -> bool:
+        r"""Dot test.
+    
+        Generate random vectors :math:`\mathbf{u}` and :math:`\mathbf{v}`
+        and perform dot-test to verify the validity of forward and adjoint
+        operators. This test can help to detect errors in the operator
+        implementation.
+    
+        Parameters
+        ----------
+        Op : LinearOperator
+            Linear operator to test
+        rtol : float, optional
+            Relative dottest tolerance
+        atol : float, optional
+            Absolute dottest tolerance
+        complexflag : bool, optional
+            Generate random vectors with complex entries for model and data.
+            Default: False
+    
+        Raises
+        ------
+        AssertionError
+            If dot-test is not verified within chosen tolerances.
+    
+        Notes
+        -----
+        A dot-test is mathematical tool used in the development of numerical
+        linear operators.
+    
+        More specifically, a correct implementation of forward and adjoint for
+        a linear operator should verify the following *equality*
+        within a numerical tolerance:
+    
+        .. math::
+            (\mathbf{Op}\,\mathbf{u})^H\mathbf{v} =
+            \mathbf{u}^H(\mathbf{Op}^H\mathbf{v})
+    
+        """
+        nr = Op.shape[0]
+        nc = Op.shape[1]
+
+        # make u and v vectors
+        rdtype = np.ones(1, Op.dtype).real.dtype
+
+        rng = np.random.default_rng()
+        u = rng.standard_normal(nc).astype(rdtype)
+
+        if complexflag:
+            u = u + 1j * rng.standard_normal(nc).astype(rdtype)
+
+        v = rng.standard_normal(nr).astype(rdtype)
+        if complexflag:
+            v = v + 1j * rng.standard_normal(nr).astype(rdtype)
+
+        op_u = Op.matvec(u)
+        opH_v = Op.rmatvec(v)
+
+        # check operator equivalence
+        assert_allclose(op_u, Op * u)
```

Comment:
not sure this makes sense as part of the dot test method... I would test this on its own...